File: wpimath/src/test/java/org/wpilib/math/kinematics/SwerveModuleVelocityTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;
import org.wpilib.math.geometry.Rotation2d;

class SwerveModuleVelocityTest {
  private static final double EPSILON = 1E-9;

  @Test
  void testOptimize() {
    var angleA = Rotation2d.fromDegrees(45);
    var refA = new SwerveModuleVelocity(-2.0, Rotation2d.PI);
    var optimizedA = refA.optimize(angleA);

    assertAll(
        () -> assertEquals(2.0, optimizedA.velocity, EPSILON),
        () -> assertEquals(0.0, optimizedA.angle.getDegrees(), EPSILON));

    var angleB = Rotation2d.fromDegrees(-50);
    var refB = new SwerveModuleVelocity(4.7, Rotation2d.fromDegrees(41));
    var optimizedB = refB.optimize(angleB);

    assertAll(
        () -> assertEquals(-4.7, optimizedB.velocity, EPSILON),
        () -> assertEquals(-139.0, optimizedB.angle.getDegrees(), EPSILON));
  }

  @Test
  void testNoOptimize() {
    var angleA = Rotation2d.ZERO;
    var refA = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(89));
    var optimizedA = refA.optimize(angleA);

    assertAll(
        () -> assertEquals(2.0, optimizedA.velocity, EPSILON),
        () -> assertEquals(89.0, optimizedA.angle.getDegrees(), EPSILON));

    var angleB = Rotation2d.ZERO;
    var refB = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(-2));
    var optimizedB = refB.optimize(angleB);

    assertAll(
        () -> assertEquals(-2.0, optimizedB.velocity, EPSILON),
        () -> assertEquals(-2.0, optimizedB.angle.getDegrees(), EPSILON));
  }

  @Test
  void testCosineScale() {
    var angleA = Rotation2d.fromDegrees(0.0);
    var refA = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedA = refA.cosineScale(angleA);

    assertAll(
        () -> assertEquals(Math.sqrt(2.0), optimizedA.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedA.angle.getDegrees(), EPSILON));

    var angleB = Rotation2d.fromDegrees(45.0);
    var refB = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedB = refB.cosineScale(angleB);

    assertAll(
        () -> assertEquals(2.0, optimizedB.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedB.angle.getDegrees(), EPSILON));

    var angleC = Rotation2d.fromDegrees(-45.0);
    var refC = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedC = refC.cosineScale(angleC);

    assertAll(
        () -> assertEquals(0.0, optimizedC.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedC.angle.getDegrees(), EPSILON));

    var angleD = Rotation2d.fromDegrees(135.0);
    var refD = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedD = refD.cosineScale(angleD);

    assertAll(
        () -> assertEquals(0.0, optimizedD.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedD.angle.getDegrees(), EPSILON));

    var angleE = Rotation2d.fromDegrees(-135.0);
    var refE = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedE = refE.cosineScale(angleE);

    assertAll(
        () -> assertEquals(-2.0, optimizedE.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedE.angle.getDegrees(), EPSILON));

    var angleF = Rotation2d.fromDegrees(180.0);
    var refF = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(45.0));
    var optimizedF = refF.cosineScale(angleF);

    assertAll(
        () -> assertEquals(-Math.sqrt(2.0), optimizedF.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedF.angle.getDegrees(), EPSILON));

    var angleG = Rotation2d.fromDegrees(0.0);
    var refG = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedG = refG.cosineScale(angleG);

    assertAll(
        () -> assertEquals(-Math.sqrt(2.0), optimizedG.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedG.angle.getDegrees(), EPSILON));

    var angleH = Rotation2d.fromDegrees(45.0);
    var refH = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedH = refH.cosineScale(angleH);

    assertAll(
        () -> assertEquals(-2.0, optimizedH.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedH.angle.getDegrees(), EPSILON));

    var angleI = Rotation2d.fromDegrees(-45.0);
    var refI = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedI = refI.cosineScale(angleI);

    assertAll(
        () -> assertEquals(0.0, optimizedI.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedI.angle.getDegrees(), EPSILON));

    var angleJ = Rotation2d.fromDegrees(135.0);
    var refJ = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedJ = refJ.cosineScale(angleJ);

    assertAll(
        () -> assertEquals(0.0, optimizedJ.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedJ.angle.getDegrees(), EPSILON));

    var angleK = Rotation2d.fromDegrees(-135.0);
    var refK = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedK = refK.cosineScale(angleK);

    assertAll(
        () -> assertEquals(2.0, optimizedK.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedK.angle.getDegrees(), EPSILON));

    var angleL = Rotation2d.fromDegrees(180.0);
    var refL = new SwerveModuleVelocity(-2.0, Rotation2d.fromDegrees(45.0));
    var optimizedL = refL.cosineScale(angleL);

    assertAll(
        () -> assertEquals(Math.sqrt(2.0), optimizedL.velocity, EPSILON),
        () -> assertEquals(45.0, optimizedL.angle.getDegrees(), EPSILON));
  }

  @Test
  void testInterpolate() {
    // Test basic interpolation with simple angles
    final var start = new SwerveModuleVelocity(1.0, Rotation2d.fromDegrees(0.0));
    final var end = new SwerveModuleVelocity(5.0, Rotation2d.fromDegrees(90.0));

    // Test interpolation at t=0 (should return start)
    final var atStart = start.interpolate(end, 0.0);
    assertAll(
        () -> assertEquals(1.0, atStart.velocity, EPSILON),
        () -> assertEquals(0.0, atStart.angle.getDegrees(), EPSILON));

    // Test interpolation at t=1 (should return end)
    final var atEnd = start.interpolate(end, 1.0);
    assertAll(
        () -> assertEquals(5.0, atEnd.velocity, EPSILON),
        () -> assertEquals(90.0, atEnd.angle.getDegrees(), EPSILON));

    // Test interpolation at t=0.5 (should return midpoint)
    final var atMidpoint = start.interpolate(end, 0.5);
    assertAll(
        () -> assertEquals(3.0, atMidpoint.velocity, EPSILON),
        () -> assertEquals(45.0, atMidpoint.angle.getDegrees(), EPSILON));

    // Test interpolation at t=0.25
    final var atQuarter = start.interpolate(end, 0.25);
    assertAll(
        () -> assertEquals(2.0, atQuarter.velocity, EPSILON),
        () -> assertEquals(22.5, atQuarter.angle.getDegrees(), EPSILON));

    // Test clamping: t < 0 should clamp to 0
    final var belowRange = start.interpolate(end, -0.5);
    assertAll(
        () -> assertEquals(1.0, belowRange.velocity, EPSILON),
        () -> assertEquals(0.0, belowRange.angle.getDegrees(), EPSILON));

    // Test clamping: t > 1 should clamp to 1
    final var aboveRange = start.interpolate(end, 1.5);
    assertAll(
        () -> assertEquals(5.0, aboveRange.velocity, EPSILON),
        () -> assertEquals(90.0, aboveRange.angle.getDegrees(), EPSILON));

    // Test angle wrapping with crossing 180/-180 boundary
    final var startWrap = new SwerveModuleVelocity(2.0, Rotation2d.fromDegrees(170.0));
    final var endWrap = new SwerveModuleVelocity(4.0, Rotation2d.fromDegrees(-170.0));
    final var midpointWrap = startWrap.interpolate(endWrap, 0.5);
    assertAll(
        () -> assertEquals(3.0, midpointWrap.velocity, EPSILON),
        () -> assertEquals(180.0, Math.abs(midpointWrap.angle.getDegrees()), EPSILON));
  }
}

File: wpimath/src/test/native/cpp/kinematics/SwerveModuleVelocityTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/kinematics/SwerveModuleVelocity.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include "wpi/math/TestAssertions.hpp"
#include "wpi/math/geometry/Rotation2d.hpp"

static constexpr double kEpsilon = 1E-9;

TEST_CASE("SwerveModuleVelocityTest Optimize", "[wpimath]") {
  wpi::math::Rotation2d angleA{45_deg};
  wpi::math::SwerveModuleVelocity refA{-2_mps, 180_deg};
  auto optimizedA = refA.Optimize(angleA);

  CHECK_NEAR(optimizedA.velocity.value(), 2.0, kEpsilon);
  CHECK_NEAR(optimizedA.angle.Degrees().value(), 0.0, kEpsilon);

  wpi::math::Rotation2d angleB{-50_deg};
  wpi::math::SwerveModuleVelocity refB{4.7_mps, 41_deg};
  auto optimizedB = refB.Optimize(angleB);

  CHECK_NEAR(optimizedB.velocity.value(), -4.7, kEpsilon);
  CHECK_NEAR(optimizedB.angle.Degrees().value(), -139.0, kEpsilon);
}

TEST_CASE("SwerveModuleVelocityTest NoOptimize", "[wpimath]") {
  wpi::math::Rotation2d angleA{0_deg};
  wpi::math::SwerveModuleVelocity refA{2_mps, 89_deg};
  auto optimizedA = refA.Optimize(angleA);

  CHECK_NEAR(optimizedA.velocity.value(), 2.0, kEpsilon);
  CHECK_NEAR(optimizedA.angle.Degrees().value(), 89.0, kEpsilon);

  wpi::math::Rotation2d angleB{0_deg};
  wpi::math::SwerveModuleVelocity refB{-2_mps, -2_deg};
  auto optimizedB = refB.Optimize(angleB);

  CHECK_NEAR(optimizedB.velocity.value(), -2.0, kEpsilon);
  CHECK_NEAR(optimizedB.angle.Degrees().value(), -2.0, kEpsilon);
}

TEST_CASE("SwerveModuleVelocityTest CosineScaling", "[wpimath]") {
  wpi::math::Rotation2d angleA{0_deg};
  wpi::math::SwerveModuleVelocity refA{2_mps, 45_deg};
  auto optimizedA = refA.CosineScale(angleA);

  CHECK_NEAR(optimizedA.velocity.value(), std::sqrt(2.0), kEpsilon);
  CHECK_NEAR(optimizedA.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleB{45_deg};
  wpi::math::SwerveModuleVelocity refB{2_mps, 45_deg};
  auto optimizedB = refB.CosineScale(angleB);

  CHECK_NEAR(optimizedB.velocity.value(), 2.0, kEpsilon);
  CHECK_NEAR(optimizedB.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleC{-45_deg};
  wpi::math::SwerveModuleVelocity refC{2_mps, 45_deg};
  auto optimizedC = refC.CosineScale(angleC);

  CHECK_NEAR(optimizedC.velocity.value(), 0.0, kEpsilon);
  CHECK_NEAR(optimizedC.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleD{135_deg};
  wpi::math::SwerveModuleVelocity refD{2_mps, 45_deg};
  auto optimizedD = refD.CosineScale(angleD);

  CHECK_NEAR(optimizedD.velocity.value(), 0.0, kEpsilon);
  CHECK_NEAR(optimizedD.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleE{-135_deg};
  wpi::math::SwerveModuleVelocity refE{2_mps, 45_deg};
  auto optimizedE = refE.CosineScale(angleE);

  CHECK_NEAR(optimizedE.velocity.value(), -2.0, kEpsilon);
  CHECK_NEAR(optimizedE.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleF{180_deg};
  wpi::math::SwerveModuleVelocity refF{2_mps, 45_deg};
  auto optimizedF = refF.CosineScale(angleF);

  CHECK_NEAR(optimizedF.velocity.value(), -std::sqrt(2.0), kEpsilon);
  CHECK_NEAR(optimizedF.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleG{0_deg};
  wpi::math::SwerveModuleVelocity refG{-2_mps, 45_deg};
  auto optimizedG = refG.CosineScale(angleG);

  CHECK_NEAR(optimizedG.velocity.value(), -std::sqrt(2.0), kEpsilon);
  CHECK_NEAR(optimizedG.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleH{45_deg};
  wpi::math::SwerveModuleVelocity refH{-2_mps, 45_deg};
  auto optimizedH = refH.CosineScale(angleH);

  CHECK_NEAR(optimizedH.velocity.value(), -2.0, kEpsilon);
  CHECK_NEAR(optimizedH.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleI{-45_deg};
  wpi::math::SwerveModuleVelocity refI{-2_mps, 45_deg};
  auto optimizedI = refI.CosineScale(angleI);

  CHECK_NEAR(optimizedI.velocity.value(), 0.0, kEpsilon);
  CHECK_NEAR(optimizedI.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleJ{135_deg};
  wpi::math::SwerveModuleVelocity refJ{-2_mps, 45_deg};
  auto optimizedJ = refJ.CosineScale(angleJ);

  CHECK_NEAR(optimizedJ.velocity.value(), 0.0, kEpsilon);
  CHECK_NEAR(optimizedJ.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleK{-135_deg};
  wpi::math::SwerveModuleVelocity refK{-2_mps, 45_deg};
  auto optimizedK = refK.CosineScale(angleK);

  CHECK_NEAR(optimizedK.velocity.value(), 2.0, kEpsilon);
  CHECK_NEAR(optimizedK.angle.Degrees().value(), 45.0, kEpsilon);

  wpi::math::Rotation2d angleL{180_deg};
  wpi::math::SwerveModuleVelocity refL{-2_mps, 45_deg};
  auto optimizedL = refL.CosineScale(angleL);

  CHECK_NEAR(optimizedL.velocity.value(), std::sqrt(2.0), kEpsilon);
  CHECK_NEAR(optimizedL.angle.Degrees().value(), 45.0, kEpsilon);
}

TEST_CASE("SwerveModuleVelocityTest Equality", "[wpimath]") {
  wpi::math::SwerveModuleVelocity state1{2_mps, 90_deg};
  wpi::math::SwerveModuleVelocity state2{2_mps, 90_deg};

  CHECK(state1 == state2);
}

TEST_CASE("SwerveModuleVelocityTest Inequality", "[wpimath]") {
  wpi::math::SwerveModuleVelocity state1{1_mps, 90_deg};
  wpi::math::SwerveModuleVelocity state2{2_mps, 90_deg};
  wpi::math::SwerveModuleVelocity state3{1_mps, 89_deg};

  CHECK(state1 != state2);
  CHECK(state1 != state3);
}
